### DISCUSSION TITLE
ZBBridge with wireguard client might became a remote secure WIFI-ZBBridge for home automation hubs

### DISCUSSION THREAD

**Discussion by urechi** [arendst/Tasmota#13789] [Ideas]
Hey all,

Wouldn't be nice to be able to have a remote Zigbee bridge connected to your Home Assistant instance and at the same time secure too? With no need to expose your MQTT broker also!
It seems that those guys have managed to get wireguard working on a ESP32 board [here](https://github.com/esphome/feature-requests/issues/1444)
How about a tasmota-zbbridge binary with the wireguard client?
We could use the socket://{wireguard_ip}:8888 for ZHA or zigbee2mqtt.
Home Assistant might even be able to discover WIFI smart devices within the LAN so that zigbee bridge became a wifi bridge also. How cool is this?

Anyway many thanks to all the contributors of Tasmota so far, you are doing a great job. Keep it going and all the best!

**Comment by s-hadinger** [comment]
Not sure to understand the use case here. 

Do you mean adding Wireguard server to the TCP bridge so that ZHA can remotely connect?

The opposite could make sense but is completely different: Tasmota connecting to a Mqtt broker through a Wireguard vpn, I.e. acting as a Wireguard client. 

Is it number 1?

**Reply by urechi** [reply]
I am thinking more like a site-to-site vpn, to allow additional clients on the same private subnet as the connecting client (tasmota) to reach the private network of the Wireguard server (home assistant for example). This way the tasmota zigbee bridge can help home assistant to discover devices connected on the same subnet but also ZHA or Z2M can connect remotely. This option would be ideal.
That being said even as a simple client, where tasmota connects to a mqtt server through wireguard, I can see the advantage because people won’t need to expose mqtt brokers to the internet anymore and communication will be through a secure tunnel. And still as a simple client I believe TCP 8888 on tasmota will be available to ZHA or Z2M, isn’t it?

**Comment by Hedda** [comment]
FYI, ZHA  (or Zigbee2MQTT/Z2M) serial connection/communication with a remote Zigbee serial adapter/module is only a good idea if used locally over a stable LAN (Local Area Network) like wired Ethernet at the same location. It is not even recommended over a Wi-Fi connection since dropped connections and high latency cause issues with remote serial communication in the serial protocols they use.

Note that there are warning in both ZHA and Zigbee2MQTT documentation about using a serial connection to Zigbee modules over Wi-Fi, including information that the reason for not doing so is poor robustness in the serial protocol that the Zigbee stacks use.

https://www.home-assistant.io/integrations/zha#warning-about-wi-fi-based-zigbee-to-serial-bridgesgateways

https://www.zigbee2mqtt.io/advanced/remote-adapter/connect_to_a_remote_adapter.html#_1-install-ser2net

ZHA (or Zigbee2MQTT/Z2M) serial connection/communication with a remote Tasmota (or ESPHome) based Zigbee network adapter over the internet via a VPN tunnel is as such not a good example use case scenario or a great idea in the real world.

A better real-world use case scenario for a remote Zigbee adapter via Tasmota (or ESPHome) is when you have Home Assistant installed in a virtual machine on a server or NAS in your basement or corner and want to have the Zigbee coordinator installed in the centre of your house on the main floor.

ZHA (and Zigbee2MQTT) really needs a stable/robust connection to the Zigbee serial module because the serial protocols that Zigbee stacks use for communication are by themselves are not made to handle loads of communications errors due to bad connections.

If you do have a secondary location, such as a holiday property (also I wish I had one), and want to use ZHA then you really need to set up an additional host with a Home Assistant instance on that location so can connect the Zigbee serial module locally directly or over the local LAN.

PS: Also note that both ZHA and Zigbee2MQTT can currently only connect to one Zigbee Coordinator for each instance and unless you implement hack you can currently also not install more than one instance of ZHA and Zigbee2MQTT in Home Assistant. Thus if you want to use two Zigbee Coordinators with ZHA you need to use more than one instance of Home Assistant. Zigbee2MQTT is a little easier since stand-alone from Home Assistant so can just install several instances of Zigbee2MQTT and then connect them all via MQTT instead.

**Comment by sevrinrobstad** [comment]
seems like there is a ESP32 wireguard implementation done her:
https://github.com/ciniml/WireGuard-ESP32-Arduino

If we somehow could run wireguard client either in ZBBridge, or better directly on all tasmota devices, we could safely control external devices.

**Comment by s-hadinger** [comment]
Is this a real need? I'm still struggling what kind of use-case this would serve.

**Reply by sevrinrobstad** [reply]
Absolutely. Not only for ZBBridge, but for Tasmota devices in general. 

The use case is to easily add Tasmota's remotely that in an easy, encrypted (and therefore safe) way can  "dial home" to a control center to be controlled.
The other ways to do this involves either cloud stuff, or involving the "remote" router in some way (VPN).
